config/config_test.go

```diff
@@ -341,6 +585,74 @@ OUTER:
 	}
 }
 
+func TestConfig_ApplyDefaults(t *testing.T) {
+	params := map[string]Parameter{
+		"limit":        {Type: ParameterTypeInt, Default: "1"},
+		"foo.*.param1": {Type: ParameterTypeString, Default: "foo"},
+		"foo.*.param2": {Type: ParameterTypeString},
+	}
+
+	testCases := []struct {
+		name string
+		have Config
+		want Config
+	}{{
+		name: "empty",
+		have: Config{},
+		want: Config{
+			"limit": "1",
+		},
+	}, {
+		name: "foo.0.param2",
+		have: Config{
+			"foo.0.param2": "bar",
+		},
+		want: Config{
+			"limit":        "1",
+			"foo.0.param1": "foo",
+			"foo.0.param2": "bar",
+		},
+	}, {
+		name: "foo.0.param1",
+		have: Config{
+			"limit":        "-1",
+			"foo.0.param1": "custom",
+		},
+		want: Config{
+			"limit":        "-1",
+			"foo.0.param1": "custom",
+			"foo.0.param2": "",
+		},
+	}, {
+		name: "multiple dynamic params",
+		have: Config{
+			"limit":                "-1",
+			"foo.0.param1":         "parameter",
+			"foo.1.param2":         "custom",
+			"foo.2.does-not-exist": "unrecognized key still triggers creation of defaults",
```

Comment:
should we add a warning for the user for this case?

Reply:
This will actually produce an error when we check for unrecognized keys (see `TestConfig_Validate_Unrecognized`). In this test we're just interested in the defaults for `foo.*.param1` and `foo.*.param2`.